File: client/rs/src/bcp.rs
use crate::types::{Error, Message};
use std::io::{Read, Write};

pub const HEADER_LEN: usize = 8;
pub type MessageCode = u16;
const SECTION_SIZE: u16 = 256;

pub const MT_ENDPOINT: MessageCode = 2;
pub const MT_INSPECT: MessageCode = 3;

pub const MT_HANDLE_INSPECT: MessageCode = (1 * SECTION_SIZE) + 0;
pub const MT_HANDLE_DROP: MessageCode = (1 * SECTION_SIZE) + 1;
pub const MT_HANDLE_KEEP_ALIVE: MessageCode = (1 * SECTION_SIZE) + 2;
pub const MT_HANDLE_SHARE_OUT: MessageCode = (1 * SECTION_SIZE) + 3;
pub const MT_HANDLE_SHARE_IN: MessageCode = (1 * SECTION_SIZE) + 4;

pub const MT_VOLUME_INSPECT: MessageCode = (2 * SECTION_SIZE) + 0;
pub const MT_VOLUME_BEGIN_TX: MessageCode = (2 * SECTION_SIZE) + 1;
pub const MT_OPEN_FIAT: MessageCode = 4;
pub const MT_OPEN_FROM: MessageCode = (2 * SECTION_SIZE) + 2;
pub const MT_CREATE_VOLUME: MessageCode = (3 * SECTION_SIZE) - 1;

pub const MT_TX_INSPECT: MessageCode = (3 * SECTION_SIZE) + 0;
pub const MT_TX_ABORT: MessageCode = (3 * SECTION_SIZE) + 1;
pub const MT_TX_COMMIT: MessageCode = (3 * SECTION_SIZE) + 2;
pub const MT_TX_LOAD: MessageCode = (3 * SECTION_SIZE) + 3;
pub const MT_TX_SAVE: MessageCode = (3 * SECTION_SIZE) + 4;
pub const MT_TX_POST: MessageCode = (3 * SECTION_SIZE) + 5;
pub const MT_TX_POST_SALT: MessageCode = (3 * SECTION_SIZE) + 6;
pub const MT_TX_GET: MessageCode = (3 * SECTION_SIZE) + 7;
pub const MT_TX_GET_SALT: MessageCode = (3 * SECTION_SIZE) + 8;
pub const MT_TX_EXISTS: MessageCode = (3 * SECTION_SIZE) + 9;
pub const MT_TX_DELETE: MessageCode = (3 * SECTION_SIZE) + 10;
pub const MT_TX_COPY: MessageCode = (3 * SECTION_SIZE) + 11;
pub const MT_TX_LINK: MessageCode = (3 * SECTION_SIZE) + 12;
pub const MT_TX_UNLINK: MessageCode = (3 * SECTION_SIZE) + 13;
pub const MT_TX_VISIT: MessageCode = (3 * SECTION_SIZE) + 14;
pub const MT_TX_IS_VISITED: MessageCode = (3 * SECTION_SIZE) + 15;
pub const MT_TX_VISIT_LINKS: MessageCode = (3 * SECTION_SIZE) + 16;

pub const MT_QUEUE_INSPECT: MessageCode = (4 * SECTION_SIZE) + 0;
pub const MT_QUEUE_ENQUEUE: MessageCode = (4 * SECTION_SIZE) + 1;
pub const MT_QUEUE_DEQUEUE: MessageCode = (4 * SECTION_SIZE) + 2;
pub const MT_QUEUE_SUB_TO_VOLUME: MessageCode = (4 * SECTION_SIZE) + 3;
pub const MT_QUEUE_CREATE: MessageCode = (5 * SECTION_SIZE) - 1;

pub const MT_OK: MessageCode = (255 * SECTION_SIZE) + 0;
pub const MT_ERROR_TIMEOUT: MessageCode = (255 * SECTION_SIZE) + 1;
pub const MT_ERROR_INVALID_HANDLE: MessageCode = (255 * SECTION_SIZE) + 2;
pub const MT_ERROR_NOT_FOUND: MessageCode = (255 * SECTION_SIZE) + 3;
pub const MT_ERROR_NO_PERMISSION: MessageCode = (255 * SECTION_SIZE) + 4;
pub const MT_ERROR_NO_LINK: MessageCode = (255 * SECTION_SIZE) + 5;
pub const MT_ERROR_TOO_LARGE: MessageCode = (255 * SECTION_SIZE) + 6;
pub const MT_ERROR_UNKNOWN: MessageCode = u16::MAX;

pub fn write_message<W: Write>(mut w: W, code: MessageCode, body: &[u8]) -> Result<(), Error> {
    let mut header = [0u8; HEADER_LEN];
    header[0] = (code >> 8) as u8;
    header[1] = (code & 0xff) as u8;
    let body_len = (body.len() as u32).to_le_bytes();
    header[4..8].copy_from_slice(&body_len);
    w.write_all(&header)?;
    w.write_all(body)?;
    Ok(())
}

pub fn read_message<R: Read>(mut r: R) -> Result<(MessageCode, Vec<u8>), Error> {
    let mut header = [0u8; HEADER_LEN];
    r.read_exact(&mut header)?;
    let code = u16::from_be_bytes([header[0], header[1]]);
    let body_len = u32::from_le_bytes([header[4], header[5], header[6], header[7]]) as usize;
    let mut body = vec![0u8; body_len];
    r.read_exact(&mut body)?;
    Ok((code, body))
}

pub fn check_ok(code: MessageCode, body: &[u8]) -> Result<(), Error> {
    if code == MT_OK {
        return Ok(());
    }
    if code > MT_OK {
        let msg = String::from_utf8_lossy(body).to_string();
        return Err(match code {
            MT_ERROR_TIMEOUT => Error::WireError { code, message: msg },
            MT_ERROR_INVALID_HANDLE => Error::WireInvalidHandle(msg),
            MT_ERROR_NOT_FOUND => Error::WireNotFound(msg),
            MT_ERROR_NO_PERMISSION => Error::WireNoPermission(msg),
            MT_ERROR_NO_LINK => Error::WireNoLink(msg),
            MT_ERROR_TOO_LARGE => Error::WireTooLarge(msg),
            MT_ERROR_UNKNOWN => Error::WireUnknown(msg),
            _ => Error::WireError { code, message: msg },
        });
    }
    Err(Error::WireError {
        code,
        message: "unexpected non-OK response code".to_string(),
    })
}

pub fn write_uvarint(out: &mut Vec<u8>, mut x: u64) {
    while x >= 0x80 {
        out.push((x as u8) | 0x80);
        x >>= 7;
    }
    out.push(x as u8);
}

pub fn read_uvarint(data: &[u8]) -> Result<(u64, &[u8]), Error> {
    let mut x = 0u64;
    let mut s = 0u32;
    for (i, b) in data.iter().enumerate() {
        if *b < 0x80 {
            if i > 9 || (i == 9 && *b > 1) {
                return Err(Error::InvalidMessage("uvarint overflow".to_string()));
            }
            return Ok((x | ((*b as u64) << s), &data[i + 1..]));
        }
        x |= ((*b & 0x7f) as u64) << s;
        s += 7;
    }
    Err(Error::InvalidMessage("truncated uvarint".to_string()))
}

pub fn append_lp(out: &mut Vec<u8>, data: &[u8]) {
    write_uvarint(out, data.len() as u64);
    out.extend_from_slice(data);
}

pub fn read_lp(data: &[u8]) -> Result<(&[u8], &[u8]), Error> {
    let (len, rest) = read_uvarint(data)?;
    let len = len as usize;
    if rest.len() < len {
        return Err(Error::InvalidMessage("truncated lp payload".to_string()));
    }
    Ok((&rest[..len], &rest[len..]))
}

pub fn decode_bools(data: &[u8], n: usize) -> Vec<bool> {
    let mut out = Vec::with_capacity(n);
    for byte in data {
        for bit in 0..8 {
            if out.len() == n {
                return out;
            }
            out.push((byte & (1 << bit)) != 0);
        }
    }
    out.resize(n, false);
    out
}

pub fn encode_message(msg: &Message) -> Vec<u8> {
    let mut out = Vec::new();
    msg.marshal(&mut out);
    out
}

File: client/zig/bcp.zig
const std = @import("std");

pub const MessageCode = u16;

pub const header_len: usize = 8;
const section_size: u16 = 256;

pub const mt_endpoint: MessageCode = 2;
pub const mt_inspect: MessageCode = 3;

pub const mt_handle_inspect: MessageCode = (1 * section_size) + 0;
pub const mt_handle_drop: MessageCode = (1 * section_size) + 1;
pub const mt_handle_keep_alive: MessageCode = (1 * section_size) + 2;
pub const mt_handle_share_out: MessageCode = (1 * section_size) + 3;
pub const mt_handle_share_in: MessageCode = (1 * section_size) + 4;

pub const mt_volume_inspect: MessageCode = (2 * section_size) + 0;
pub const mt_volume_begin_tx: MessageCode = (2 * section_size) + 1;
pub const mt_open_fiat: MessageCode = 4;
pub const mt_open_from: MessageCode = (2 * section_size) + 2;
pub const mt_create_volume: MessageCode = (3 * section_size) - 1;

pub const mt_tx_inspect: MessageCode = (3 * section_size) + 0;
pub const mt_tx_abort: MessageCode = (3 * section_size) + 1;
pub const mt_tx_commit: MessageCode = (3 * section_size) + 2;
pub const mt_tx_load: MessageCode = (3 * section_size) + 3;
pub const mt_tx_save: MessageCode = (3 * section_size) + 4;
pub const mt_tx_post: MessageCode = (3 * section_size) + 5;
pub const mt_tx_post_salt: MessageCode = (3 * section_size) + 6;
pub const mt_tx_get: MessageCode = (3 * section_size) + 7;
pub const mt_tx_get_salt: MessageCode = (3 * section_size) + 8;
pub const mt_tx_exists: MessageCode = (3 * section_size) + 9;
pub const mt_tx_delete: MessageCode = (3 * section_size) + 10;
pub const mt_tx_copy: MessageCode = (3 * section_size) + 11;
pub const mt_tx_link: MessageCode = (3 * section_size) + 12;
pub const mt_tx_unlink: MessageCode = (3 * section_size) + 13;
pub const mt_tx_visit: MessageCode = (3 * section_size) + 14;
pub const mt_tx_is_visited: MessageCode = (3 * section_size) + 15;
pub const mt_tx_visit_links: MessageCode = (3 * section_size) + 16;

pub const mt_queue_inspect: MessageCode = (4 * section_size) + 0;
pub const mt_queue_enqueue: MessageCode = (4 * section_size) + 1;
pub const mt_queue_dequeue: MessageCode = (4 * section_size) + 2;
pub const mt_queue_sub_to_volume: MessageCode = (4 * section_size) + 3;
pub const mt_queue_create: MessageCode = (5 * section_size) - 1;

pub const mt_ok: MessageCode = (255 * section_size) + 0;
pub const mt_error_timeout: MessageCode = (255 * section_size) + 1;
pub const mt_error_invalid_handle: MessageCode = (255 * section_size) + 2;
pub const mt_error_not_found: MessageCode = (255 * section_size) + 3;
pub const mt_error_no_permission: MessageCode = (255 * section_size) + 4;
pub const mt_error_no_link: MessageCode = (255 * section_size) + 5;
pub const mt_error_too_large: MessageCode = (255 * section_size) + 6;
pub const mt_error_unknown: MessageCode = std.math.maxInt(u16);

pub const Response = struct {
    code: MessageCode,
    body: []u8,
};

pub fn appendInt(out: *std.array_list.Managed(u8), comptime T: type, value: T, endian: std.builtin.Endian) !void {
    var buf: [@sizeOf(T)]u8 = undefined;
    std.mem.writeInt(T, &buf, value, endian);
    try out.appendSlice(&buf);
}

pub fn writeMessage(writer: anytype, code: MessageCode, body: []const u8) !void {
    var header = std.mem.zeroes([header_len]u8);
    header[0] = @truncate(code >> 8);
    header[1] = @truncate(code & 0xff);
    std.mem.writeInt(u32, header[4..8], @intCast(body.len), .little);
    try writer.writeAll(&header);
    try writer.writeAll(body);
}

fn readExact(reader: anytype, out: []u8) !void {
    const reader_ty = @TypeOf(reader.*);
    if (@hasDecl(reader_ty, "readSliceAll")) {
        try reader.readSliceAll(out);
        return;
    }
    if (@hasDecl(reader_ty, "readNoEof")) {
        try reader.readNoEof(out);
        return;
    }
    @compileError("reader type must support readSliceAll or readNoEof");
}

pub fn readMessageAlloc(allocator: std.mem.Allocator, reader: anytype) !Response {
    var header: [header_len]u8 = undefined;
    try readExact(reader, &header);
    const code = (@as(u16, header[0]) << 8) | @as(u16, header[1]);
    const body_len = std.mem.readInt(u32, header[4..8], .little);
    const body = try allocator.alloc(u8, body_len);
    errdefer allocator.free(body);
    try readExact(reader, body);
    return .{ .code = code, .body = body };
}

pub fn checkOk(code: MessageCode) !void {
    if (code == mt_ok) return;
    if (code > mt_ok) {
        return switch (code) {
            mt_error_timeout => error.WireTimeout,
            mt_error_invalid_handle => error.WireInvalidHandle,
            mt_error_not_found => error.WireNotFound,
            mt_error_no_permission => error.WireNoPermission,
            mt_error_no_link => error.WireNoLink,
            mt_error_too_large => error.WireTooLarge,
            mt_error_unknown => error.WireUnknown,
            else => error.WireError,
        };
    }
    return error.UnexpectedResponseCode;
}

pub fn writeUvarint(out: *std.array_list.Managed(u8), value: u64) !void {
    var x = value;
    while (x >= 0x80) {
        try out.append(@as(u8, @truncate(x)) | 0x80);
        x >>= 7;
    }
    try out.append(@truncate(x));
}

pub fn readUvarint(data: []const u8) !struct { value: u64, rest: []const u8 } {
    var x: u64 = 0;
    var shift: u6 = 0;
    for (data, 0..) |b, i| {
        if (b < 0x80) {
            if (i > 9 or (i == 9 and b > 1)) return error.InvalidMessage;
            return .{ .value = x | (@as(u64, b) << shift), .rest = data[i + 1 ..] };
        }
        x |= (@as(u64, b & 0x7f) << shift);
        shift += 7;
    }
    return error.InvalidMessage;
}

pub fn appendLp(out: *std.array_list.Managed(u8), payload: []const u8) !void {
    try writeUvarint(out, payload.len);
    try out.appendSlice(payload);
}

pub fn readLp(data: []const u8) !struct { payload: []const u8, rest: []const u8 } {
    const len_and_rest = try readUvarint(data);
    if (len_and_rest.rest.len < len_and_rest.value) return error.InvalidMessage;
    const n: usize = @intCast(len_and_rest.value);
    return .{ .payload = len_and_rest.rest[0..n], .rest = len_and_rest.rest[n..] };
}

pub fn decodeBoolBitset(allocator: std.mem.Allocator, bytes: []const u8, n: usize) ![]bool {
    const out = try allocator.alloc(bool, n);
    var i: usize = 0;
    for (bytes) |byte| {
        var bit: u3 = 0;
        while (bit < 8) : (bit += 1) {
            if (i == n) return out;
            out[i] = (byte & (@as(u8, 1) << bit)) != 0;
            i += 1;
        }
    }
    while (i < n) : (i += 1) out[i] = false;
    return out;
}

test "uvarint roundtrip" {
    var da = std.heap.DebugAllocator(.{}){};
    defer _ = da.deinit();
    const allocator = da.allocator();

    var buf = std.array_list.Managed(u8).init(allocator);
    defer buf.deinit();

    try writeUvarint(&buf, 0);
    try writeUvarint(&buf, 127);
    try writeUvarint(&buf, 128);
    try writeUvarint(&buf, 65_535);

    var rest: []const u8 = buf.items;
    for ([_]u64{ 0, 127, 128, 65_535 }) |want| {
        const got = try readUvarint(rest);
        try std.testing.expectEqual(want, got.value);
        rest = got.rest;
    }
    try std.testing.expectEqual(@as(usize, 0), rest.len);
}
